packages/jest-runner/src/types.ts

```diff
@@ -37,12 +41,45 @@ export type OnTestSuccess = (
   testResult: TestResult,
 ) => Promise<void>;
 
+// Typings for `sendMessageToJest` events
+
+export type TestFileStart = {
+  eventName: 'test-file-start';
+  args: [Test];
+};
+export type TestFileSuccess = {
+  eventName: 'test-file-success';
+  args: [Test, TestResult];
+};
+export type TestFileFailure = {
+  eventName: 'test-file-failure';
+  args: [Test, SerializableError];
+};
+export type TestFileResult = {
+  eventName: 'test-case-result';
+  args: [Config.Path, AssertionResult];
+};
+
+export type TestFileEvent = (
+  eventName:
+    | TestFileStart['eventName']
+    | TestFileSuccess['eventName']
+    | TestFileFailure['eventName']
+    | TestFileResult['eventName'],
+  args:
+    | TestFileStart['args']
+    | TestFileSuccess['args']
+    | TestFileFailure['args']
+    | TestFileResult['args'],
+) => unknown;
```

Comment:
```suggestion
export type TestFileEvent =
    | TestFileStart
    | TestFileSuccess
    | TestFileFailure
    | TestFileResult;
```

Reply:
I'm happy to play witht his and the above if it turns out to be finicky

Reply:
![image](https://user-images.githubusercontent.com/28642011/88340112-d6a25380-cd58-11ea-9c76-689e074125a9.png)

I'm just keeping it like before, coz the new typings are causing the params to be `never` as the types in the union has multiple overloads.

I'm adding a simple reproduce-able case here:

```typescript
type F1 = (
  a: string, b: number
) => void;

type F2 = (
  a: string, b: boolean
) => void;


type F3 = F1 | F2;
const f: F3 = (a: Parameters<F3>[0], b: Parameters<F3>[1]) => {
  console.log(a, b);
};

f('apple', 3); // Error: Argument of type '3' is not assignable to parameter of type 'never'.(2345)
```

[Playground Link](https://www.staging-typescript.org/play?#code/C4TwDgpgBAYgjFAvFAFAWAFBSgQwFxQDOwATgJYB2A5gDRQBGBFArgLb0QmYCUSAfFABuAezIATANyZMoSLABMSVJmz4ipSrQYF6w4QBsIOCj35DRk6RhnhoMAMxL4UAD4KpGAMbCKxKADMCByUUNQAFHBIcVghgTkIAHgc+AG0ABgBdOkYoCKiYuJJE5JS4DN5EAQBvFShvXwMIADp9YSpQ7O4PAF8PTH8UAHIcMDBDQbp7LqA)

A good explanation of the same issue is provided in [typescript changelogs](https://github.com/Microsoft/TypeScript/wiki/What%27s-new-in-TypeScript#improved-behavior-for-calling-union-types)